milli/src/update/index_documents/transform.rs

```diff
@@ -340,35 +340,45 @@ impl<'a, 'i> Transform<'a, 'i> {
             return Ok(None);
         }
 
+        // store the keys and values of the json + the original obkv
+        let mut key_value: Vec<(FieldId, Cow<[u8]>)> = Vec::new();
+
+        // the object we're going to use to
         let mut doc = serde_json::Map::new();
 
-        for (k, v) in obkv.iter() {
-            let key = self.fields_ids_map.name(k).ok_or(FieldIdMapMissingEntry::FieldId {
-                field_id: k,
-                process: "Flatten from fields ids map.",
-            })?;
-            let value = serde_json::from_slice::<serde_json::Value>(v)
-                .map_err(crate::error::InternalError::SerdeJson)?;
-            doc.insert(key.to_string(), value);
+        // we recreate a json containing only the fields that needs to be flattened.
+        // all the raw values get inserted directly in the `key_value` vec.
+        for (key, value) in obkv.iter() {
+            if json_depth_checker::should_flatten_from_unchecked_slice(value) {
+                let key = self.fields_ids_map.name(key).ok_or(FieldIdMapMissingEntry::FieldId {
+                    field_id: key,
+                    process: "Flatten from fields ids map.",
+                })?;
+
+                let value = serde_json::from_slice::<serde_json::Value>(value)
+                    .map_err(crate::error::InternalError::SerdeJson)?;
+                doc.insert(key.to_string(), value);
+            } else {
+                key_value.push((key, value.into()));
+            }
         }
 
         let flattened = flatten_serde_json::flatten(&doc);
 
-        // Once we have the flattened version we can convert it back to obkv and
-        // insert all the new generated fields_ids (if any) in the fields ids map.
-        let mut buffer: Vec<u8> = Vec::new();
-        let mut writer = KvWriter::new(&mut buffer);
-        let mut flattened: Vec<_> = flattened.into_iter().collect();
-        // we reorder the field to get all the known field first
-        flattened
-            .sort_unstable_by_key(|(key, _)| self.fields_ids_map.id(&key).unwrap_or(FieldId::MAX));
-
-        for (key, value) in flattened {
+        // Once we have the flattened version we insert all the new generated fields_ids
+        // (if any) in the fields ids map and serialize the value.
+        for (key, value) in flattened.into_iter() {
             let fid = self.fields_ids_map.insert(&key).ok_or(UserError::AttributeLimitReached)?;
             let value = serde_json::to_vec(&value).map_err(InternalError::SerdeJson)?;
-            writer.insert(fid, &value)?;
+            key_value.push((fid, value.into()));
         }
 
+        // we sort the key. If there was a conflict between the obkv and the new generated value the
+        // keys will be consecutive.
+        key_value.sort_unstable_by(|(a, _), (b, _)| a.cmp(b));
```

Comment:
```suggestion
        key_value.sort_unstable_by_key(|(fid, _)| *fid);
```

Reply:
I can't do that;
```
error: lifetime may not live long enough
   --> milli/src/update/index_documents/transform.rs:378:49
    |
378 |         key_value.sort_unstable_by_key(|(a, _)| a);
    |                                         ------- ^ returning this value requires that `'1` must outlive `'2`
    |                                         |     |
    |                                         |     return type of closure is &'2 u16
    |                                         has type `&'1 (u16, Cow<'_, [u8]>)`
```

Reply:
But can't you just deref this `u16` (`FieldId`), here?

Reply:
Ah yes you're right!